### PR TITLE
LLM: Adapt transformers models for `optimize model` SL

### PR DESCRIPTION
**Pull request by leonardozcm**
## Description

In order to adapt the from_pretrained and save/load API for different Transformer models, adapt more torch load involved methods `_load_from_state_dict `

## Usage Example
```python
#####First load using `from_pretrained` and call `optimize_model`
from transformers import AutoModelForCausalLM
from bigdl.llm import optimize_model

path = '/disk1/changmin/moss-moon-003-sft'
saved_dir = '/disk1/changmin/moss-moon-003-sft-INT4'

model = AutoModelForCausalLM.from_pretrained(path,
                                  torch_dtype="auto",
                                  low_cpu_mem_usage=True,
                                  trust_remote_code=True)

model = optimize_model(model)

model.save_low_bit(saved_dir) # For transformer models we directly call `save_pretrained`

#####Use `load_low_bit` to load with torch model on cpu
from bigdl.llm.optimize import load_low_bit

model = AutoModelForCausalLM.from_pretrained(path,
                                  torch_dtype="auto",
                                  low_cpu_mem_usage=True,
                                  trust_remote_code=True)

model = load_low_bit(model, saved_dir)

#####Use `load_low_bit` to load with torch model on meta device
from bigdl.llm.optimize import low_memory_init, load_low_bit

with low_memory_init():
     model = AutoModelForCausalLM.from_pretrained(path, torch_dtype="auto", trust_remote_code=True)

model = load_low_bit(model, saved_dir)
```

## Further Test
1. chatglm2-6b(4G of int4 model)
image

2. llama2-7b(3.8G ...)
image

3. bloom-7b(7.4G ...)
image

4. Baichuan 13B(7.7G ...)
```log
--------------------before model init--------------------

Used: 10027.57 MB
--------------------after model init--------------------

Used: 10032.98 MB
--------------------after model load--------------------

Used: 10031.48 MB
--------------------after model load--------------------

Used: 17858.24 MB
```

5. (Shards model) moss(9.4G+1.2G)
image

### CODE REVIEW

**Comment by leonardozcm** [comment]
### Usage1: Traditionally load in-memory model on cpu:
```python
    from bigdl.llm.optimize import load_low_bit
    path = '/disk1/changmin/chatglm2-6b'
    saved_dir = '/disk1/changmin/chatglm2-bigdl-llm-INT4'
   
    # From original model path
    model = AutoModel.from_pretrained(path, # from original model dir
                                      torch_dtype="auto",
                                      low_cpu_mem_usage=True,
                                      trust_remote_code=True)
   
    # From saved model path
    model = AutoModel.from_pretrained(saved_dir,  # from saved model dir
                                      ignore_mismatched_sizes = True, # need this to avoid weights miss match error
                                      torch_dtype="auto",
                                      low_cpu_mem_usage=True,
                                      trust_remote_code=True)
    model = load_low_bit(model,
                        model_path=saved_dir)
```

For Usage1, note that:
1. You need both` original transformer model path` and saved `lowbit model path`. The original model path contains configs that helps transformer build the fp32 model, and then we convert and load it with our lowbit weights. Or you can add `ignore_mismatched_sizes = True` to load from saved model.
2. `low_cpu_mem_usage` is optional, but recommended.

### Usage2: Fast and low cost load model on meta device:
```python
    from bigdl.llm.optimize import low_memory_init, load_low_bit
    # path = '/disk1/changmin/chatglm2-6b'
    saved_dir = '/disk1/changmin/chatglm2-bigdl-llm-INT4'

    with low_memory_init():
        model = AutoModel.from_pretrained(saved_dir , torch_dtype="auto", trust_remote_code=True)

    model = load_low_bit(model,
                        model_path=saved_dir)
```

For Usage2, note that:
1. The model is fully loaded only after load_low_bit; you are not expect to use it before that.
2. No need to pass `low_cpu_mem_usage=True`
3. Different from  Usage1, we don't need `ignore_mismatched_sizes = True,` and ``low_cpu_mem_usage = True`, but it influence nothing to keep them.

### Sanity Check:
chatglm2-6b, both usage.1 and usage.2
```log
-------------------- Prompt --------------------
问：AI是什么？

答：
-------------------- Output --------------------
问：AI是什么？

答： AI指的是人工智能,是一种能够通过学习和推理来执行任务的计算机程序。它可以模仿人类的思维方式,做出类似人类的决策,并且具有自主学习、自我
```

File: python/llm/src/bigdl/llm/utils/lazy_load_torch.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+             torch.int8: 1}
+
+@dataclass
+class LazyStorage:
+    load: Callable[[int, int], torch.Tensor]
+    kind: StorageType
+    description: str
+
+@dataclass
+class LazyTensor:
```
Comment (leonardozcm) [inline python/llm/src/bigdl/llm/utils/lazy_load_torch.py:67]: We introduce `LazyTensor` here to make our loading method smarter, for in openai.whisper, they save not only Tensors but also some model configs like:
```python
    with (
        io.BytesIO(checkpoint_file) if in_memory else open(checkpoint_file, "rb")
    ) as fp:
        checkpoint = torch.load(fp, map_location=device)
    del checkpoint_file

    dims = ModelDimensions(**checkpoint["dims"])
```
This requires an actual read operation on the serialized file. With lazytensor we can read other objects as normal, but read no huge tensors to memory in advance,

Comment (leonardozcm) [reply python/llm/src/bigdl/llm/utils/lazy_load_torch.py:70]: I can't make `openai.whisper` 1*memory size load for they force read model file with python built-in `open(..).read()` to check the model's check sum
https://github.com/openai/whisper/blob/main/whisper/__init__.py#L58-L59

**Comment by leonardozcm** [comment]
For transformer models:

1. chatglm2-6b, 4G

image

2. llama2-7b, 3.8G
 
image

3. bloom, 7G

image

**Review by plusbang — APPROVED** [review]

**Comment by jason-dai** [comment]
Do we have test for this?

**Comment by leonardozcm** [comment]
> Do we have test for this?

will add some uts before merge

**Comment by leonardozcm** [comment]
image

**Pull request merged**